### ISSUE TITLE
[BUG] go-micro/v3 http_transport Recv and Close data race on buff

### ISSUE THREAD

**Issue by bbdshow** [go-micro/go-micro#2431]
**Describe the bug**
V3 http_transport has Data Race, too much traffic can easily trigger panic 

Hope v3 version to get ' http_transport Recv and Close Data Race on Buff ' #2374 fix

The #2374 fix is currently fixed in v4.5.0, but we can't upgrade to v4 at the moment

**Comment by bbdshow** [comment]
Or is there another way to solve the problem? 😊

**Comment by xpunch** [comment]
Have you tested the solution merged in #2374 can solve your issue?

**Comment by bbdshow** [comment]
> 

There is no test, I was just checking the V3.6.0 problem, I found the exception here, it is really the data race caused by http_transport, see the official Merge to V4.5.0, I only have this basis, because the cause of Panic condition is very strict, so I do not have a test

**Comment by xpunch** [comment]
If you can re-produce your issue, can your help confirm whether it can solve your issue?
As you can see in #2374,  the real problem is already merged in #419, I can help you merge those changes into v3, but it'll be more helpful if you're sure it really solve your issue.

**Comment by bbdshow** [comment]
Ok, I will test V3.6.0 and V4.5.0 through the method of stress test. To reproduce the problem and determine whether V4.5.0 is solved  😊

**Comment by bbdshow** [comment]
hello, I tested it today, v3.6.0 100% recurrence is to make rpc_client timeout.

eg:
client.NewClient(client.RequestTimeout(X*time.Second), client.Registry(etcd.NewRegistry(registry.Addrs(etcdAddr...))))

If the rpc_service processing time exceeds Xs and there is concurrency, a data race will occur.

When I use the same condition settings, v4.5.0 can respond to 408 code Timeout normally, and will not panic

Please also merge #2374 into V3 if you have time. Thank you!

Or you can follow the example above to reproduce the v3.6.0 client Panic

**Comment by xpunch** [comment]
Already merged into v3.7.1.

**Comment by umishra21** [comment]
---
"Context deadline exceeded error "
---

I upgraded from v1 to v3 of go-micro in my application but i am getting context deadline exceeded when receiving parallel request. Due to this, the memory runs out and services crash. For small number, the application is working for let say 5-6 simultaneous request but fails for around 200 or more requests.

Go Version: go version go1.13 
```
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/ec2-user/.cache/go-build"
GOENV="/home/ec2-user/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/ec2-user/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build260645297=/tmp/go-build -gno-record-gcc-switches"
```
